pyomo/contrib/pynumero/interfaces/pyomo_grey_box_nlp.py

```diff
@@ -0,0 +1,770 @@
+
+        finally:
+            # Restore the ctypes of the ExternalGreyBoxBlock components
+            for greybox in greybox_components:
+                greybox.parent_block().reclassify_component_type(
+                    greybox, ExternalGreyBoxBlock)
+
+        if self._pyomo_nlp.n_primals() == 0:
+            raise ValueError(
+                "No variables were found in the Pyomo part of the model."
+                " PyomoGreyBoxModel requires at least one variable"
+                " to be active in a Pyomo objective or constraint")
+
+        # build the list of NLP wrappers for the greybox objects
+        greybox_nlps = []
+        fixed_vars = []
+        for greybox in greybox_components:
+            # iterate through the data objects if component is indexed
+            for data in greybox.values():
+                if data.active:
+                    # check that no variables are fixed
+                    fixed_vars.extend(v for v in data.inputs.values() if v.fixed)
+                    fixed_vars.extend(v for v in data.outputs.values() if v.fixed)
+                    greybox_nlp = _ExternalGreyBoxAsNLP(greybox)
+                    greybox_nlps.append(greybox_nlp)
+
+        if fixed_vars:
+            logging.getLogger(__name__).error('PyomoNLPWithGreyBoxBlocks found fixed variables for the'
+                                              ' inputs and/or outputs of an ExternalGreyBoxBlock. This'
+                                              ' is not currently supported. The fixed variables were:\n\t'
+                                              + '\n\t'.join(f.getname(fully_qualified=True) for f in fixed_vars)
+                                              )
+            raise NotImplementedError('PyomoNLPWithGreyBoxBlocks does not support fixed inputs or outputs')
+
+        # let's build up the union of all the primal variables names
+        primals_names = set(self._pyomo_nlp.primals_names())
+        for gbnlp in greybox_nlps:
+            primals_names.update(gbnlp.primals_names())
+
+        # sort the names for consistency run to run
+        self._n_primals = len(primals_names)
+        self._primals_names = primals_names = sorted(primals_names)
+        self._pyomo_model_var_datas = [self._pyomo_model_var_names_to_datas[nm] for nm in self._primals_names]
+
+        # get the names of all the constraints
+        self._constraint_names = list(self._pyomo_nlp.constraint_names())
+        self._constraint_datas = [self._pyomo_model_constraint_names_to_datas.get(nm) for nm in self._constraint_names]
+        for gbnlp in greybox_nlps:
+            self._constraint_names.extend(gbnlp.constraint_names())
+            self._constraint_datas.extend([(gbnlp._block, nm) for nm in gbnlp.constraint_names()])
+        self._n_constraints = len(self._constraint_names)
+
+        self._hessian_support = True
+        for nlp in greybox_nlps:
+            if not nlp.hessian_support():
```

Comment:
Maybe too late, but could we consider renaminig this method?  Maybe `has_hessian_support` or `supports_hessian_computation` or even `supports_hessians`?

Reply:
agreed - changed to has_hessian_support.